packages/platforms/mellanox/x86-64/x86-64-mlnx-msn2740/onlp/builds/src/module/src/fani.c

```diff
@@ -0,0 +1,450 @@
+    char min[LEN_FILE_NAME];
+    char max[LEN_FILE_NAME];
+}fan_path_T;
+
+#define _MAKE_FAN_PATH_ON_MAIN_BOARD(prj,id) \
+    { #prj"fan"#id"_status", \
+      #prj"fan"#id"_speed_get", \
+      #prj"fan"#id"_speed_set", \
+      #prj"fan"#id"_min", \
+      #prj"fan"#id"_max" }
+
+#define MAKE_FAN_PATH_ON_MAIN_BOARD(prj,id) _MAKE_FAN_PATH_ON_MAIN_BOARD(prj,id)
+
+#define MAKE_FAN_PATH_ON_PSU(psu_id, fan_id) \
+    {"psu"#psu_id"_status", \
+     "psu"#psu_id"_fan"#fan_id"_speed_get", "", "", "",}
+
+static fan_path_T fan_path[] =  /* must map with onlp_fan_id */
+{
+    MAKE_FAN_PATH_ON_MAIN_BOARD(PROJECT_NAME, FAN_RESERVED),
+    MAKE_FAN_PATH_ON_MAIN_BOARD(PROJECT_NAME, FAN_1_ON_MAIN_BOARD),
+    MAKE_FAN_PATH_ON_MAIN_BOARD(PROJECT_NAME, FAN_2_ON_MAIN_BOARD),
+    MAKE_FAN_PATH_ON_MAIN_BOARD(PROJECT_NAME, FAN_3_ON_MAIN_BOARD),
+    MAKE_FAN_PATH_ON_MAIN_BOARD(PROJECT_NAME, FAN_4_ON_MAIN_BOARD),
+    MAKE_FAN_PATH_ON_PSU(1 ,1),
+    MAKE_FAN_PATH_ON_PSU(2, 1)
+};
+
+#define MAKE_FAN_INFO_NODE_ON_MAIN_BOARD(id) \
+    { \
+        { ONLP_FAN_ID_CREATE(FAN_##id##_ON_MAIN_BOARD), "Chassis Fan "#id, 0 }, \
+        0x0, \
+        (ONLP_FAN_CAPS_SET_PERCENTAGE | ONLP_FAN_CAPS_GET_PERCENTAGE | \
+         ONLP_FAN_CAPS_GET_RPM | ONLP_FAN_CAPS_SET_RPM), \
+        0, \
+        0, \
+        ONLP_FAN_MODE_INVALID, \
+    }
+
+#define MAKE_FAN_INFO_NODE_ON_PSU(psu_id, fan_id) \
+    { \
+        { ONLP_FAN_ID_CREATE(FAN_##fan_id##_ON_PSU##psu_id), "Chassis PSU-"#psu_id" Fan "#fan_id, 0 }, \
+        0x0, \
+        (ONLP_FAN_CAPS_GET_RPM | ONLP_FAN_CAPS_GET_PERCENTAGE), \
+        0, \
+        0, \
+        ONLP_FAN_MODE_INVALID, \
+    }
+
+/* Static fan information */
+onlp_fan_info_t linfo[] = {
+    { }, /* Not used */
+    MAKE_FAN_INFO_NODE_ON_MAIN_BOARD(1),
+    MAKE_FAN_INFO_NODE_ON_MAIN_BOARD(2),
+    MAKE_FAN_INFO_NODE_ON_MAIN_BOARD(3),
+    MAKE_FAN_INFO_NODE_ON_MAIN_BOARD(4),
+    MAKE_FAN_INFO_NODE_ON_PSU(1,1),
+    MAKE_FAN_INFO_NODE_ON_PSU(2,1)
+};
+
+#define VALIDATE(_id)                           \
+    do {                                        \
+        if(!ONLP_OID_IS_FAN(_id)) {             \
+            return ONLP_STATUS_E_INVALID;       \
+        }                                       \
+    } while(0)
+
+#define OPEN_READ_FILE(fullpath, data, nbytes, len)			\
+	if (onlp_file_read((uint8_t*)data, nbytes, &len, fullpath) < 0)	\
+       return ONLP_STATUS_E_INTERNAL;           \
+	else													\
+		AIM_LOG_VERBOSE("read data: %s\n", r_data);			\
+
+
+/* No support in _onlp_fani_read_fan_eepron() function as System Fans does not contain eeprom. */
+
+static int
+_onlp_fani_info_get_fan(int local_id, onlp_fan_info_t* info)
+{
+    int   len = 0, nbytes = 10;
+    float range = 0;
+    float temp  = 0;
+    char  r_data[10]   = {0};
+    char  fullpath[65] = {0};
+
+    /* get fan status
+    */
+    snprintf(fullpath, sizeof(fullpath), "%s%s", PREFIX_MODULE_PATH, fan_path[local_id].status);
```

Comment:
I understand this is a copied pattern from earlier implementations. 

Please note that the onlp_file_* APIs already take a format string and variable arguments for generating the filename. Can you replace the snprintf/OPEN_READ_FILE combos here with just a native call to the onlp_file_* APIs?

Reply:
pulling back request and issuing new one with comments.